pallets/ajuna-affiliate/src/lib.rs

```diff
@@ -0,0 +1,272 @@
+	/// Such accounts can't be affiliatees anymore.
+	#[pallet::storage]
+	#[pallet::getter(fn affiliators)]
+	pub type Affiliators<T: Config> =
+		StorageMap<_, Identity, T::AccountId, AffiliatorState, ValueQuery>;
+
+	/// Stores the affiliate logic rules
+	#[pallet::storage]
+	pub type AffiliateRules<T: Config> =
+		StorageMap<_, Blake2_128Concat, ExtrinsicId, u8, OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An organizer has been set.
+		OrganizerSet {
+			organizer: T::AccountId,
+		},
+		AccountAffiliated {
+			account: T::AccountId,
+			to: T::AccountId,
+		},
+		RuleAdded {
+			extrinsic_id: ExtrinsicId,
+		},
+		RuleCleared {
+			extrinsic_id: ExtrinsicId,
+		},
+		RuleExecuted {
+			extrinsic_id: ExtrinsicId,
+			account: T::AccountId,
+			beneficiary: T::AccountId,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// There is no account set as the organizer
+		OrganizerNotSet,
+		/// An account cannot affiliate itself
+		CannotAffiliateSelf,
+		/// The account is not allowed to receive affiliates
+		CannotAffiliateToAccount,
+		/// This account has reached the affiliate limit
+		CannotAffiliateMoreAccounts,
+		/// This account has already been affiliated by another affiliator
+		CannotAffiliateAlreadyAffiliatedAccount,
+		/// This account is already an affiliator, so it cannot affiliate to another account
+		CannotAffiliateExistingAffiliator,
+		/// The account is blocked, so it cannot be affiliated to
+		CannotAffiliateBlocked,
+		/// The given extrinsic identifier is already paired with an affiliate rule
+		ExtrinsicAlreadyHasRule,
+		/// The given extrinsic identifier doesn't have any rule associated with it
+		MissingRuleForExtrinsic,
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn add_new_affiliate_to(
+			affiliator: T::AccountId,
+			affiliatee: T::AccountId,
+		) -> DispatchResult {
+			let accounts = {
+				let mut accounts_vec = match Affiliatees::<T>::take(&affiliator) {
+					Some(accounts) => accounts,
+					None => AffiliatedAccountsOf::<T>::new(),
+				};
+
+				Self::try_add_account_to(&mut accounts_vec, affiliator.clone())?;
+
+				accounts_vec
+			};
+
+			Affiliatees::<T>::insert(affiliatee, accounts);
+			Affiliators::<T>::mutate(&affiliator, |state| {
+				state.affiliates = state.affiliates.saturating_add(1);
+			});
+
+			Ok(())
+		}
+
+		fn try_add_account_to(
+			accounts: &mut AffiliatedAccountsOf<T>,
+			account: T::AccountId,
+		) -> DispatchResult {
+			if accounts.len() == T::AffiliateMaxLevel::get() as usize {
+				accounts.pop();
+			}
+			accounts
+				.try_insert(0, account)
+				.map_err(|_| Error::<T>::CannotAffiliateMoreAccounts.into())
+		}
+	}
+
+	impl<T: Config> AffiliateInspector<AccountIdOf<T>> for Pallet<T> {
+		fn get_affiliator_chain_for(account: &AccountIdOf<T>) -> Option<Vec<AccountIdOf<T>>> {
+			Affiliatees::<T>::get(account).map(|accounts| accounts.to_vec())
```

Comment:
`StorageMap::get()` returns owned values, but `to_vec` performs unfortunately `clone()`, so we can save `O(n)` if we replace `accounts.to_vec()` with `accounts.into_inner()`.

Reply:
Oh, didn't know that! I'll make the change!